### DISCUSSION TITLE
DPD messages are sent to peer , in spite of ingress traffic received in IPv6

### DISCUSSION THREAD

**Discussion by shyampujar** [strongswan/strongswan#1779] [Q&A]
1. DPD messages are not  expected to be sent to peer, if there is ingress traffic. ( indicating peer is alive ) 
   https://wiki.strongswan.org/issues/2376
2. Same is also observed for ipv4.
3. However there is deviation in IPv6.
4. DPD messages are not sent, only if both side (ingress && egress) packets are seen.

Could you please comment, why do we have separate behavior for ipv4 and ipv6? 
is there any configuration to be done, so that only ingress traffic is considered for DPD?

Strongswan version used: 5.7.2 
Linux distro: Centos 7.9

Regards,
Shyam

**Comment by tobiasbrunner** [comment]
Are you sure you are tunneling your IPv6 traffic?

**Reply by shyampujar** [reply]
Yes I am tunneling Ipv6 traffic. I will share the PCAP file regarding the same, in sometime.

**Reply by shyampujar** [reply]
![DPD-issue](https://github.com/strongswan/strongswan/assets/11404522/238a0d43-99d5-40c2-952f-863fc667dd8b)

Attached setup-details for reference

**Comment by shyampujar** [comment]
I have found the difference between working and non-working setup. Unable to upload  the image due to policy restriction.
 
Local: fd03:298c:8dfe:825::10b:5001
Remote: fd03:298c:8dfe:828::26:1001
TSi: 192.168.10.32/29 
TSr: 192.168.1.111/32 

Could you please comment as

1. What is the need of checking Inner IP subnet ( IPv4, TSi ) presence on the system , with IPv6 ( Local) Tunnel ?
    If i add the IP 192.168.10.43/28 on lo device ( which covers ip range 192.168.10.32-47) , then DPD is not triggered. 
   If i change the IP addition to 192.168.10.43/29 ( range: 192.168.10.40-47) , then DPD is always triggered.

2. why IPv4 - inner IP subnet presence ( TSi) , does not matter for IPv4( Local ) tunnel?

**Reply by Kentzo** [reply]
Does the behavior change if you [disable MOBIKE](https://docs.strongswan.org/docs/5.9/features/mobike.html#:~:text=strongSwan%20is%20running%20the%20MOBIKE,conf%20.)?

**Reply by shyampujar** [reply]
mobike=no is already set in swanctl.conf in both peers
As per the our need, we cannot enable mobike in our deployment.

**Reply by tobiasbrunner** [reply]
DPDs are sent based on last (inbound) use time of the IKE SA and IPsec policies (or IPsec SAs on newer kernels). This does not require any local IPs, nor does MOBIKE have any effect on it. You can check the status output (in strongSwan and via `ip -s xfrm policy`).

**Reply by shyampujar** [reply]
Following is the snapshot of tcpdump and policy stats details:
==================
13:47:33.998348 IP6 fe80::887f:c3ff:feef:e0e0 > ff02::1: ICMP6, router advertisement, length 32
13:47:34.169349 IP6 fe80::f8af:3cff:fe16:62b1 > ff02::1: ICMP6, router advertisement, length 32
13:47:34.822057 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:47:36.805210 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x13), length 116
13:47:36.805210 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:47:37.828658 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:47:38.571052 IP6 fe80::49d:27ff:febd:4065 > ff02::1: ICMP6, router advertisement, length 32
13:47:40.832050 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:47:43.238828 IP6 fe80::7c47:5fff:fea0:1f38 > ff02::1: ICMP6, router advertisement, length 32
13:47:43.809494 IP6 fe80::880:75ff:fe8a:256d > ff02::1: ICMP6, router advertisement, length 32
^C
30 packets captured
30 packets received by filter
0 packets dropped by kernel
4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# ip -s xfrm policy
src 192.168.10.32/29 dst 192.168.1.111/32 uid 0
        dir out action allow index 55537 priority 368767 ptype main share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-07-11 13:46:03 use -
        tmpl src fd03:298c:8dfe:825::10b:4001 dst fd03:298c:8dfe:828::26:1001
                proto esp spi 0xc949f2b4(3377066676) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 192.168.1.111/32 dst 192.168.10.32/29 uid 0
        dir fwd action allow index 55530 priority 368767 ptype main share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-07-11 13:46:03 use -
        tmpl src fd03:298c:8dfe:828::26:1001 dst fd03:298c:8dfe:825::10b:4001
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 192.168.1.111/32 dst 192.168.10.32/29 uid 0
        dir in action allow index 55520 priority 368767 ptype main share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-07-11 13:46:03 use -
        tmpl src fd03:298c:8dfe:828::26:1001 dst fd03:298c:8dfe:825::10b:4001
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode tunnel
                level required share any
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff

===============================

4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# swanctl -l
Ike1: #1, ESTABLISHED, IKEv2, 3e3047f1afec5da2_i* aa30d85a884294b7_r
  local  'O=Altiostar, CN=ca.altiostar.com' @ fd03:298c:8dfe:825::10b:4001[500]
  remote 'O=Altiostar, CN=ca.altiostar.com' @ fd03:298c:8dfe:828::26:1001[500]
  AES_CBC-128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
  established 277s ago, rekeying in 37540s, reauth in 70557s
  Child1: #1, reqid 1, INSTALLED, TUNNEL, ESP:AES_CBC-128/HMAC_SHA1_96
    installed 277s ago, rekeying in 6612s, expires in 7643s
    in  cdd307ea,   1330 bytes,    19 packets
    out c949f2b4,      0 bytes,     0 packets
    local  192.168.10.32/29
    remote 192.168.1.111/32

**Reply by shyampujar** [reply]
tcpdump  with TSi subnet not present:  DPD is always triggered. 
dpd_delay=30s

4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# ip addr add 192.168.10.43/29 dev lo
4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# tcpdump -i vnet -nn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on vnet, link-type RAW (Raw IP), capture size 262144 bytes
13:53:33.567519 IP6 fe80::48bd:afff:fee1:11a9 > ff02::1: ICMP6, router advertisement, length 32
13:53:33.805357 IP6 fd03:298c:8dfe:825::10b:4001.500 > fd03:298c:8dfe:828::26:1001.500: isakmp: child_sa  inf2[I]
13:53:33.805888 IP6 fd03:298c:8dfe:828::26:1001.500 > fd03:298c:8dfe:825::10b:4001.500: isakmp: child_sa  inf2[R]
13:53:36.271342 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64

13:53:36.805101 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x5b), length 116
13:53:36.805101 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:53:36.822575 IP6 fe80::880:75ff:fe8a:256d > ff02::1: ICMP6, router advertisement, length 32

13:53:38.623748 IP6 fe80::c85a:37ff:fe12:7caf > ff02::1: ICMP6, router advertisement, length 32
13:53:40.155087 IP6 fe80::f4e9:89ff:fe6a:7199 > ff02::1: ICMP6, router advertisement, length 32
13:53:40.278144 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:53:44.281374 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:53:46.147202 IP6 fe80::34b6:e6ff:fe95:6e9b > ff02::1: ICMP6, router advertisement, length 32
13:53:47.287975 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:53:50.294934 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:53:51.805095 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x5e), length 116
13:53:51.805095 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:53:53.301431 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:53:54.004685 IP6 fe80::887f:c3ff:feef:e0e0 > ff02::1: ICMP6, router advertisement, length 32
13:53:55.930500 IP6 fe80::2872:2fff:fe6c:94f8 > ff02::1: ICMP6, router advertisement, length 32
13:53:57.308271 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:01.314865 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:02.426929 IP6 fe80::6423:b3ff:fefc:a3bb > ff02::1: ICMP6, router advertisement, length 32
13:54:03.805620 IP6 fd03:298c:8dfe:825::10b:4001.500 > fd03:298c:8dfe:828::26:1001.500: isakmp: child_sa  inf2[I]
13:54:03.806139 IP6 fd03:298c:8dfe:828::26:1001.500 > fd03:298c:8dfe:825::10b:4001.500: isakmp: child_sa  inf2[R]
13:54:04.321516 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64

**Reply by shyampujar** [reply]
tcpdump and TSi subnet present on the system.
DPD is not triggered. 
==================
4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# ip addr del 192.168.10.43/29 dev lo
4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# ip addr add 192.168.10.43/28 dev lo
4g-du-pod-sp-enodeb-561-du-10-pod-0:/root# tcpdump -i vnet -nn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on vnet, link-type RAW (Raw IP), capture size 262144 bytes
13:54:28.242998 IP6 fe80::482f:66ff:fe13:54d0 > ff02::1: ICMP6, router advertisement, length 32
13:54:28.348212 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:29.892881 IP6 fe80::14df:38ff:fe2f:3655 > ff02::1: ICMP6, router advertisement, length 32
13:54:30.233443 IP6 fe80::e05a:80ff:feb4:7702 > ff02::1: ICMP6, router advertisement, length 32
13:54:31.351622 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:33.805886 IP6 fd03:298c:8dfe:825::10b:4001.500 > fd03:298c:8dfe:828::26:1001.500: isakmp: child_sa  inf2[I]
13:54:33.806432 IP6 fd03:298c:8dfe:828::26:1001.500 > fd03:298c:8dfe:825::10b:4001.500: isakmp: child_sa  inf2[R]
13:54:34.355160 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:36.805118 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x67), length 116
13:54:36.805118 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:54:37.361713 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:41.368426 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:42.153310 IP6 fe80::7856:d8ff:fe74:e996 > ff02::1: ICMP6, router advertisement, length 32
13:54:44.371910 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:48.127957 IP6 fe80::3ccb:76ff:fe1a:8813 > ff02::1: ICMP6, router advertisement, length 32
13:54:48.378400 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:49.151209 IP6 fe80::c867:e2ff:fe8c:5648 > ff02::1: ICMP6, router advertisement, length 32
13:54:49.429400 IP6 fe80::346f:b0ff:fe6f:7837 > ff02::1: ICMP6, router advertisement, length 32
13:54:51.379366 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:51.806473 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x6a), length 116
13:54:51.806473 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:54:52.496843 IP6 fe80::6c14:ddff:fe9b:7306 > ff02::1: ICMP6, router advertisement, length 32
13:54:55.385125 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:54:59.391839 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:55:02.718531 IP6 fe80::e851:46ff:fe3c:30b2 > ff02::1: ICMP6, router advertisement, length 32
13:55:03.395281 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:55:04.117139 IP6 fe80::d408:15ff:fec1:8379 > ff02::1: ICMP6, router advertisement, length 32
13:55:04.218994 IP6 fe80::3832:a1ff:fe4e:9dfd > ff02::1: ICMP6, router advertisement, length 32
13:55:06.805121 IP6 fd03:298c:8dfe:828::26:1001 > fd03:298c:8dfe:825::10b:4001: ESP(spi=0xcdd307ea,seq=0x6d), length 116
13:55:06.805121 IP 192.168.1.111.50592 > 192.168.10.33.50592: UDP, length 42
13:55:07.401819 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:55:10.405286 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
13:55:10.968223 IP6 fe80::c456:a2ff:fe32:5772 > ff02::1: ICMP6, router advertisement, length 32
13:55:11.598585 IP6 fe80::55:f6ff:feb4:e949 > ff02::1: ICMP6, router advertisement, length 32
13:55:12.643847 IP6 fe80::82d:5aff:fe96:533f > ff02::1: ICMP6, router advertisement, length 32
13:55:13.369490 IP6 fe80::949e:bbff:fec7:3065 > ff02::1: ICMP6, router advertisement, length 32
13:55:13.408524 IP6 fe80::8e94:1fff:fe64:74bf > ff02::1: ICMP6, router advertisement, length 64
^C
37 packets captured
37 packets received by filter
0 packets dropped by kernel

**Reply by tobiasbrunner** [reply]
> src 192.168.1.111/32 dst 192.168.10.32/29 uid 0
> dir fwd action allow index 55530 priority 368767 ptype main share any flag (0x00000000)
> ...
> add 2023-07-11 13:46:03 use -
> ...
> src 192.168.1.111/32 dst 192.168.10.32/29 uid 0
> dir in action allow index 55520 priority 368767 ptype main share any flag (0x00000000)
> ...
> add 2023-07-11 13:46:03 use -

The kernel doesn't report a use time for the in- nor the outbound policy. While strongSwan reports traffic on the inbound SA (you can check the kernel with `ip -s xfrm state`:

> in cdd307ea, 1330 bytes, 19 packets

Note that strongSwan also doesn't report a use time (would be after the number of packets). So either the decrypted traffic is rejected due to a policy mismatch, or the kernel for some reason doesn't update the last use time of the policies (could be a bug in the kernel you are using - since it's CentOS it might be older with lots of patches).

**Reply by shyampujar** [reply]
With the above shared logs, only 2 things are of importance.
1. TSi - subnet is absent on the system. - DPD is initiated.
    this indicates, decrypted packets are rejected. cause there is no forwarding rule in system. 
   
2. TSi-subnet is present on the system. DPD is not initiated.
   packets are accepted and dropped though there is no application listening for packets. 
In my understanding, in both cases, kernel should update the last use time of policies, when there is successful decryption.
It should not depend on return status of "decrypted packet".  Just like, NIC(interface) level stats get update in/out , there are separate error counters, which indicate packets rx/tx was successful or not. 

What do you suggest , to see the expected behavior of DPD from strongswan, irrespective of TSi subnet presence on system?

**Reply by tobiasbrunner** [reply]
> this indicates, decrypted packets are rejected. cause there is no forwarding rule in system.

What do you mean? What forwarding rule?  And why would that have anything to do with the IPsec policies? Note that strongSwan checks the in and fwd policies, so forwarded traffic is seen as active traffic (as long as the kernel updates the last use time of the policy).




> In my understanding, in both cases, kernel should update the last use time of policies, when there is successful decryption.

Yep. If it does not, it's a kernel bug.

**Reply by shyampujar** [reply]
Following is the analysis:
1. Linux distro: Centos 7.9  , kernel: 3.10.0-1160.53.1.el7.x86_64
2.  Linux distro: Rocky Linux release 8.6 (Green Obsidian) , Kernel: 4.18.0-372.32.1.rt7.189.el8_6.x86_64

Only ingress traffic is present on the System. 
1. If TSi is absent on system and route to forward the decrypted packet is present, then No DPD is triggered.
2. If TSi present on the system, no DPD is triggered. 

If decrypted packet is  Self destined / forward rule  present on system, then only its considered as successful , for DPD to work.
Could you please comment, is this decision taken by charon Or Kernel updates last use time for policies only in above conditions mentioned?

Regards,
Shyam
